paymentsheet/src/main/java/com/stripe/android/paymentsheet/IntentConfirmationHandler.kt

```diff
@@ -0,0 +1,323 @@
+            is InternalPaymentResult.Failed -> Result.Failed(
+                cause = result.throwable,
+                message = resolvableString(R.string.stripe_something_went_wrong),
+                type = ErrorType.Payment,
+            )
+            is InternalPaymentResult.Canceled -> Result.Canceled
+        }
+
+        deferredIntentConfirmationType = null
+
+        completableResult?.complete(intentResult)
+
+        removeIsAwaitingForPaymentResult()
+    }
+
+    private fun onFailure(
+        cause: Throwable,
+        message: ResolvableString,
+        type: ErrorType,
+    ) {
+        completableResult?.complete(
+            Result.Failed(
+                cause = cause,
+                message = message,
+                type = type,
+            )
+        )
+    }
+
+    private fun withPaymentLauncher(action: (PaymentLauncher) -> Unit) {
+        paymentLauncher?.let(action) ?: run {
+            onFailure(
+                cause = IllegalArgumentException(
+                    "No 'PaymentLauncher' instance was created before starting confirmation. Did you call register?"
+                ),
+                message = resolvableString(R.string.stripe_something_went_wrong),
+                type = ErrorType.Fatal,
+            )
+        }
+    }
+
+    private fun storeIsAwaitingForPaymentResult() {
+        savedStateHandle[AWAITING_PAYMENT_RESULT_KEY] = true
+    }
+
+    private fun removeIsAwaitingForPaymentResult() {
+        savedStateHandle.remove<Boolean>(AWAITING_PAYMENT_RESULT_KEY)
+    }
+
+    private fun isAwaitingForPaymentResult(): Boolean {
+        return savedStateHandle.get<Boolean>(AWAITING_PAYMENT_RESULT_KEY) ?: false
+    }
+
+    internal data class Args(
+        val initializationMode: PaymentSheet.InitializationMode,
+        val shippingDetails: AddressDetails?,
+    )
+
+    /**
+     * Defines the result types that [IntentConfirmationHandler] can return after completing the confirmation process.
+     */
+    sealed interface Result {
+        /**
+         * Indicates that the confirmation process was canceled by the customer.
+         */
+        data object Canceled : Result
+
+        /**
+         * Indicates that the confirmation process has been successfully completed. A [StripeIntent] with an updated
+         * state is returned as part of the result as well.
+         */
+        data class Succeeded(
+            val intent: StripeIntent,
+            val deferredIntentConfirmationType: DeferredIntentConfirmationType?
+        ) : Result
+
+        /**
+         * Indicates that the confirmation process has failed. A cause and potentially a resolvable message are
+         * returned as part of the result.
+         */
+        data class Failed(
+            val cause: Throwable,
+            val message: ResolvableString,
+            val type: ErrorType,
+        ) : Result
+    }
+
+    enum class ErrorType {
+        Fatal,
+        Payment,
+        NextStep,
+    }
```

Comment:
I introduced these error types to better indicate where the error occurred because `PaymentSheet` deals with these errors differently
- On a fatal error, the sheet is closed with a failed result.
- On a payment result, the sheet is closed with a failed result and a payment failure event is reported.
- On a next step failure, a user error is shown in the sheet (they can retry payment)